### ISSUE TITLE
Time & Cost for this training

### ISSUE THREAD

**Issue by yhyu13** [IBM/Dromedary#3]
Hi, 

What is the approximate time & cost of training a dromedary 65B model?

You have set a time limit of 6 hours in each training step, and you are using a 64 * 6 V100GPU cluster. So the training cost will be approximately 1 day of renting a 64 * 6 V100 cluster?

I am trying to replicate a 7B LLaMa model with a dromedary pipeline.

Also, looking forward seeing your sharing the evaluation code mentioned in TODO list soon :+1:

**Comment by yhyu13** [comment]
Some follow-up questions:

1, Do you have the plan to release the generated synthetic dataset (step1 & step2) used for fine-tuning?
2, In principle, I can let GPT4 or any other existing LLM generate synthetic data for me, and then use these datasets to fine-tune a LLaMa with your step3 & step4?
3, This pipeline does not limit to fine-tuning LLaMA pre-trained models but can be generalized to self-align any LLM as long as the corresponding prompt format is adjusted to that LLM?

**Comment by Edward-Sun** [comment]
Hi Hang,

We will update the project soon after the NeurIPS submission deadline.

Regarding your question:

> 1. You have set a time limit of 6 hours in each training step, and you are using a 64 * 6 V100GPU cluster. So the training cost will be approximately 1 day of renting a 64 * 6 V100 cluster?

The 64 * 6 V100GPU are used for generating the synthetic data. The actual training (i.e., LoRA-based fine-tuning) usually only needs 16 x 6 V100GPU or less (if you use gradient accumulation).

> 1, Do you have the plan to release the generated synthetic dataset (step1 & step2) used for fine-tuning?

Yes, we will update the project soon after the NeurIPS submission deadline.

> 2, In principle, I can let GPT4 or any other existing LLM generate synthetic data for me, and then use these datasets to fine-tune a LLaMa with your step3 & step4?

AFAIK, the available GPT-4 or GPT-3.5 models released are already aligned, which means they may not be suitable **base** language models for research if you want to study the alignment of language models. Though, if you just want to get your own powerful AI assistant, [distilling from GPT-4](https://twitter.com/lupantech/status/1652022897563795456?s=61&t=jFgajP61uXtY9lrIIipivw) seems to be quite effective.

> 3, This pipeline does not limit to fine-tuning LLaMA pre-trained models but can be generalized to self-align any LLM as long as the corresponding prompt format is adjusted to that LLM?

Yes. Besides, since LLaMA is just a **base** language model, I believe the prompt format in our project can be applied to other LLMs without any modification.

**Comment by TheBloke** [comment]
> The 64 * 6 V100GPU are used for generating the synthetic data. The actual training (i.e., LoRA-based fine-tuning) usually only needs 16 x 6 V100GPU or less (if you use gradient accumulation).

@Edward-Sun 

Apologies, but I'm not quite following what you mean by '6 V100GPU'?

Does this mean the 16GB V100GPU, so 16 x 16GB V100 = 256GB VRAM total?  Or is the '6' a version number of the V100 or something like that?

Thank you very much.

**Comment by Edward-Sun** [comment]
Hi @TheBloke, 6 V100GPU means six V100-32GB GPUs, that means we used 16 x 6 = 96 V100-32GB GPUs in total (~ 3072 GB VRAM).

This is because we use a quite large batch size of 768 (i.e., per-device batch size of 8). I believe the total VRAM can be significantly reduced if you use a smaller batch size or gradient accumulations.

**Comment by TheBloke** [comment]
Oh wow! OK that's a lot more than I thought :)

But yes of course with such a large batch size, many would be needed. I didn't register that you had such a high batch size.

Thank you very much for getting back to me, that's very useful to know.

**Comment by ghost** [comment]
> Hi Hang,
> 
> We will update the project soon after the NeurIPS submission deadline.
> 
> Regarding your question:
> 
> > 1. You have set a time limit of 6 hours in each training step, and you are using a 64 * 6 V100GPU cluster. So the training cost will be approximately 1 day of renting a 64 * 6 V100 cluster?
> 
> The 64 * 6 V100GPU are used for generating the synthetic data. The actual training (i.e., LoRA-based fine-tuning) usually only needs 16 x 6 V100GPU or less (if you use gradient accumulation).
> 
> > 1, Do you have the plan to release the generated synthetic dataset (step1 & step2) used for fine-tuning?
> 
> Yes, we will update the project soon after the NeurIPS submission deadline.
> 
> > 2, In principle, I can let GPT4 or any other existing LLM generate synthetic data for me, and then use these datasets to fine-tune a LLaMa with your step3 & step4?
> 
> AFAIK, the available GPT-4 or GPT-3.5 models released are already aligned, which means they may not be suitable **base** language models for research if you want to study the alignment of language models. Though, if you just want to get your own powerful AI assistant, [distilling from GPT-4](https://twitter.com/lupantech/status/1652022897563795456?s=61&t=jFgajP61uXtY9lrIIipivw) seems to be quite effective.
> 
> > 3, This pipeline does not limit to fine-tuning LLaMA pre-trained models but can be generalized to self-align any LLM as long as the corresponding prompt format is adjusted to that LLM?
> 
> Yes. Besides, since LLaMA is just a **base** language model, I believe the prompt format in our project can be applied to other LLMs without any modification.

Could you let me know as to how much time would it take to run dromedary on llama2 7b on A100 gpu.
